### ISSUE TITLE
Discrepancy Between Reported Max Frames and Max Frame: ReadExhausted

### ISSUE THREAD

**Issue by JosephCatrambone** [oddity-ai/video-rs#50]
Hi, all.  Thanks for all the work on this library.  I've got a tool I was putting together to do a little bit of roto-ing.

I noticed for a few files the number of frames is reported as Y, but when I seek past frame X (where X < Y) I see a ReadExhausted.

```
Loaded video with size 576x576, 767 frames.
Fetching frame 0
Fetching frame 765
thread 'main' panicked at src\frame_provider\video_provider.rs:59:61:
Failed to decode frame from video stream.: ReadExhausted
```

Nore that this:

```
	let mut video = Decoder::new(file).expect("failed to create decoder");
	let (video_width, video_height) = video.size();
	let last_frame = video.frames().unwrap() as usize;
	println!("Loaded video with size {}x{}, {} frames.", &video_width, &video_height, &last_frame);
```

Reports 767 frames while the read fails after 765.

I thought it might have been an off-by-one error at first, but it happened with another video that reported 462 frames but failed after 442 frames, a difference of 20.

The full source for this is here: https://github.com/JosephCatrambone/autoro/commit/3942300485aba440d5c9fb5105d8c2e48eb7049b

The video I had used for testing is attached.  My current guess is that I'm just not using the decode() method correctly and am assuming it automatically increments the frame number.

https://github.com/user-attachments/assets/49d0cf54-cb51-4f75-93bb-dd976c57e30f

**Comment by gerwin3** [comment]
Thanks for reporting! I think it may be a bug caused by a misunderstanding on my part on how to correctly drain the ffmpeg decoder. I implemented a possible fix in #51. Can you test if the fix produces correct results? You can swap out your Cargo.toml dependency like this:

```toml
video-rs =  { git = "https://github.com/oddity-ai/video-rs", branch = "fix/drain-decoder" }
```

**Comment by JosephCatrambone** [comment]
I'll do my best to test it out tonight or tomorrow night.  In the event I forget and you don't hear back from me by Friday, please feel welcome (but not obligated) to ping me again. 😓

**Comment by JosephCatrambone** [comment]
Tried it:
```
Loaded video with size 576x576, 767 frames.
Fetching frame 0
...
Fetching frame 765
Fetching frame 766
Fetching frame 766
```

That looks like it resolves it!

I see a different but probably related issue on that branch when I seek towards the end.  If I try to fetch frame 766 of 767 after fetching 766 a few times:

```
Loaded video with size 576x576, 767 frames.
Fetching frame 0
...
Fetching frame 765
Fetching frame 766
Fetching frame 765
Failed to decode frame from video stream.: DecodeExhausted
```

The code that does this:

```
// Note this seek works okay if we don't hit the last frame.
self.video_decode_stream.seek_to_frame(target_frame as i64).expect("Unable to seek to frame.");
let (ts, frame) = self.video_decode_stream.decode().expect("Failed to decode frame from video stream.");
```

It appears that with the changes seeking back does not behave quite as well, but seek_to_start() still works as expected.

**Comment by gerwin3** [comment]
Thanks for testing! What do you mean by fetching exactly? Could you share your code?

**Comment by JosephCatrambone** [comment]
Sure.  The code and repo are in the first post, but this section in particular: https://github.com/JosephCatrambone/autoro/blob/main/src/frame_provider/video_provider.rs#L46

Seeking to a frame and decoding it.

**Comment by gerwin3** [comment]
Yeah it seems that the decoder has already drained and is basically in a "done" state. This will happen after you feed it the last frame from the input. It is really hard to work around this because decoders inherently expect a linear stream. One other problem you may have, even if you were able to seek to arbitrary points in the input file, you'd still get decoding artifacts in the video unless you seek only to keyframes.

There are two things you can try:

* Rework your code to read and decode the video file linearly and store the frames in memory. Then you can random access them afterwards just fine.
* Split the decoder and wire the reader and decoder yourself. `video_rs::Decoder` is essentially a convenience API that combines reading and decoding. But you can call `into_parts` to deconstruct it to the underlying reader and decoder. It does add some complication because you now have to correctly handle the read and decode loop yourself (including draining and flushing correctly). Take into account you can only restart arbitrary decoding on keyframes so random access to frames is quite hard. Probably the first option is the easiest ;)

**Comment by JosephCatrambone** [comment]
I have concerns about decoding the full video in memory.  If someone has raw 8k footage, that's ~121 GB/minute and I've only got 64 gigs of memory.  😅 

A compromise/workaround might be to only seek forward and cache recent frames, then seek to start if I hit the end.  *Will seek_to_start() reset the "done" state?*  If so, then I think I can just use the forward iteration as is and do something like `if next_requested_frame < last_decoded_frame { seek_to_start(); while frame < next_requested_frame { decode() } }`.

**Comment by gerwin3** [comment]
> Will seek_to_start() reset the "done" state?

No, unfortunately not.

Most likely you'll need to split the decoder as mentioned earlier. Then you'll have control over the reading process. You can just read frames and feed them to your decoder whenever you like. That will suit more advanced use cases.

It depends a bit on what you are doing, but it seems like you are building a video player. Note that this seems like a simple task, but it really isn't (I know from experience). Anyway, what you want to do in that case is first, split the decoder into reader and decoder (into_parts). Then, depending on the video player state:

* If it is in normal "playing mode", read a frame and feed it to the decoder (do this in a separate thread). Put the decoded frame in a player queue. Display the frame. The worker thread should be "greedy" as in decode frames as quickly as possible but only up to a fixed number of frames beyond the currently playing frame. So if the player is playing frame 40, the decoder should decode as fast as possible up to frame 60 for example, then stop and wait until the next frame is played, etc. This way you can guarantee smooth playback. (This is hard.)
* When you get to the last frame from the reader. Drain the decoder. There may be multiple frames. That is not a problem since you can put them all on the player queue. After that you have to restart the entire decoder. Not the reader. The reader can just seek to start.
* When the user seeks: Seek to the specified index and then either run frames backwards or forwards until you find a keyframe. When you have a keyframe, you can feed it to the decoder as if nothing happened, then continue afterwards. You don't need to reset it (I think)

**Comment by JosephCatrambone** [comment]
Many thanks!  Appreciate the input.  If things progress to the point where smooth playback is required I think I'll try exactly that.  

At this point the tool is not intended to be a video player, per se; being able to show videos is secondary to doing automatic rotoscoping and depth models with video streams. (DaVinci Resolves auto-roto is frustratingly buggy.)
